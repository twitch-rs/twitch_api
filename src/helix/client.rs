//! Client for Helix endpoints
use crate::client::ResponseExt;

use super::*;

pub(crate) mod client_ext;
#[cfg(feature = "unsupported")]
mod custom;

#[cfg(feature = "client")]
impl<C: crate::HttpClient<'static> + crate::client::ClientDefault<'static>> Default
    for HelixClient<'static, C>
{
    fn default() -> Self { Self::new() }
}

/// Client for Helix or the [New Twitch API](https://dev.twitch.tv/docs/api)
///
/// Use [`HelixClient::new`] or [`HelixClient::with_client`] to create a new client.
///
/// ```rust
/// use twitch_api::HelixClient;
/// # pub mod reqwest {pub type Client = twitch_api::client::DummyHttpClient;}
/// let helix: HelixClient<reqwest::Client> = HelixClient::new();
/// ```
///
/// See [`req_get`](HelixClient::req_get) for [`GET`](RequestGet),
/// [`req_put`](HelixClient::req_put) for [`PUT`](RequestPut),
/// [`req_post`](HelixClient::req_post) for [`POST`](RequestPost),
/// [`req_patch`](HelixClient::req_patch) for [`PATCH`](RequestPatch) and
/// [`req_delete`](HelixClient::req_delete) for [`DELETE`](RequestDelete)
///
/// Most [clients][crate::HttpClient] will be able to use the `'static` lifetime
///
/// ```rust,no_run
/// # use twitch_api::{HelixClient}; pub mod reqwest {pub type Client = twitch_api::client::DummyHttpClient;}
/// pub struct MyStruct {
///     twitch: HelixClient<'static, reqwest::Client>,
///     token: twitch_oauth2::AppAccessToken,
/// }
/// // etc
/// ```
///
/// See [`HttpClient`][crate::HttpClient] for implemented http clients, you can also define your own if needed.
///
/// # Examples
///
/// Get a [user](users::User) from their login name.
///
/// ```rust,no_run
/// # #[tokio::main]
/// # async fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
/// # pub mod reqwest {pub type Client = twitch_api::client::DummyHttpClient;}
/// use twitch_api::helix::{users::User, HelixClient};
/// let client: HelixClient<'static, reqwest::Client> = HelixClient::default();
/// # let token = twitch_oauth2::AccessToken::new("validtoken".to_string());
/// # let token = twitch_oauth2::UserToken::from_existing(&client, token, None, None).await?;
/// let user: Option<User> = client
///     .get_user_from_login("justintv", &token)
///     .await
///     .unwrap();
/// # Ok(()) }
/// ```
#[derive(Clone)]
#[cfg(all(feature = "client", feature = "helix"))] // this is needed due to a bug?
pub struct HelixClient<'a, C>
where C: crate::HttpClient<'a> {
    pub(crate) client: C,
    pub(crate) _pd: std::marker::PhantomData<&'a ()>, // TODO: Implement rate limiter...
}

#[cfg(feature = "client")]
impl<'a, C: crate::HttpClient<'a>> HelixClient<'a, C> {
    /// Create a new client with an existing client
    pub fn with_client(client: C) -> HelixClient<'a, C> {
        HelixClient {
            client,
            _pd: std::marker::PhantomData::default(),
        }
    }

    /// Create a new [`HelixClient`] with a default [`HttpClient`][crate::HttpClient]
    pub fn new() -> HelixClient<'a, C>
    where C: crate::client::ClientDefault<'a> {
        let client = C::default_client();
        HelixClient::with_client(client)
    }

    /// Retrieve a clone of the [`HttpClient`][crate::HttpClient] inside this [`HelixClient`]
    pub fn clone_client(&self) -> C
    where C: Clone {
        self.client.clone()
    }

    /// Retrieve a reference of the [`HttpClient`][crate::HttpClient] inside this [`HelixClient`]
    pub fn get_client(&self) -> &C { &self.client }

    /// Request on a valid [`RequestGet`] endpoint
    ///
    /// ```rust,no_run
    /// # #[tokio::main]
    /// # async fn main() {
    /// #   use twitch_api::helix::{HelixClient, channels};
    /// #   let token = Box::new(twitch_oauth2::UserToken::from_existing_unchecked(
    /// #       twitch_oauth2::AccessToken::new("totallyvalidtoken".to_string()), None,
    /// #       twitch_oauth2::ClientId::new("validclientid".to_string()), None, "justintv".into(), "1337".into(), None, None));
    /// let ids: &[&twitch_types::UserIdRef] = &["123456".into()];
    /// let req = channels::GetChannelInformationRequest::broadcaster_ids(ids);
    /// let client = HelixClient::new();
    /// # let _: &HelixClient<twitch_api::DummyHttpClient> = &client;
    ///
    /// let response = client.req_get(req, &token).await;
    /// # }
    /// # // fn main() {run()}
    /// ```
    pub async fn req_get<'req, R, T>(
        &'a self,
        request: R,
        token: &T,
    ) -> Result<
        Response<R, yoke::Yoke<R::Response, Vec<u8>>>,
        ClientRequestError<<C as crate::HttpClient<'a>>::Error>,
    >
    where
        R: Request + RequestGet,
        for<'y> R::Response: yoke::Yokeable<'y>,
        for<'y> yoke::trait_hack::YokeTraitHack<<R::Response as yoke::Yokeable<'y>>::Output>:
            serde::Deserialize<'y>,
        T: TwitchToken + ?Sized,
        C: Send,
    {
        let req = request.create_request(token.token().secret(), token.client_id().as_str())?;
        let uri = req.uri().clone();
        let response = self
            .client
            .req(req)
            .await
            .map_err(ClientRequestError::RequestError)?
            .into_response_vec()
            .await?;
        let (parts, body) = response.into_parts();
        let mut pagination = None;
        let mut request_opt = None;
        let mut total = None;
        let mut other = None;
        let resp: yoke::Yoke<_, _> = yoke::Yoke::try_attach_to_cart::<_, _>(
            body,
            |body| -> Result<_, ClientRequestError<<C as crate::HttpClient<'a>>::Error>> {
                let response = http::Response::from_parts(parts, body);
                let Response {
                    data,
                    pagination: pagination_inner,
                    request: request_inner,
                    total: total_inner,
                    other: other_inner,
                }: Response<_, _> = <R>::parse_response(Some(request), &uri, &response).unwrap();
                pagination = pagination_inner;
                request_opt = request_inner;
                total = total_inner;
                other = other_inner;
                Ok(data)
            },
        )?;

        Ok(Response {
            data: resp,
            pagination,
            request: request_opt,
            total,
            other,
        })
    }

    // /// Request on a valid [`RequestPost`] endpoint
    // pub async fn req_post<R, B, D, T>(
    //     &'a self,
    //     request: R,
    //     body: B,
    //     token: &T,
    // ) -> Result<Response<R, D>, ClientRequestError<<C as crate::HttpClient<'a>>::Error>>
    // where
    //     R: Request<Response = D> + Request + RequestPost<Body = B>,
    //     B: HelixRequestBody,
    //     D: serde::de::DeserializeOwned + PartialEq,
    //     T: TwitchToken + ?Sized,
    // {
    //     let req =
    //         request.create_request(body, token.token().secret(), token.client_id().as_str())?;
    //     let uri = req.uri().clone();
    //     let response = self
    //         .client
    //         .req(req)
    //         .await
    //         .map_err(ClientRequestError::RequestError)?
    //         .into_response_vec()
    //         .await?;
    //     <R>::parse_response(Some(request), &uri, response).map_err(Into::into)
    // }

    // /// Request on a valid [`RequestPatch`] endpoint
    // pub async fn req_patch<R, B, D, T>(
    //     &'a self,
    //     request: R,
    //     body: B,
    //     token: &T,
    // ) -> Result<Response<R, D>, ClientRequestError<<C as crate::HttpClient<'a>>::Error>>
    // where
    //     R: Request<Response = D> + Request + RequestPatch<Body = B>,
    //     B: HelixRequestBody,
    //     D: serde::de::DeserializeOwned + PartialEq,
    //     T: TwitchToken + ?Sized,
    // {
    //     let req =
    //         request.create_request(body, token.token().secret(), token.client_id().as_str())?;
    //     let uri = req.uri().clone();
    //     let response = self
    //         .client
    //         .req(req)
    //         .await
    //         .map_err(ClientRequestError::RequestError)?
    //         .into_response_vec()
    //         .await?;
    //     <R>::parse_response(Some(request), &uri, response).map_err(Into::into)
    // }

    // /// Request on a valid [`RequestDelete`] endpoint
    // pub async fn req_delete<R, D, T>(
    //     &'a self,
    //     request: R,
    //     token: &T,
    // ) -> Result<Response<R, D>, ClientRequestError<<C as crate::HttpClient<'a>>::Error>>
    // where
    //     R: Request<Response = D> + Request + RequestDelete,
    //     D: serde::de::DeserializeOwned + PartialEq,
    //     T: TwitchToken + ?Sized,
    // {
    //     let req = request.create_request(token.token().secret(), token.client_id().as_str())?;
    //     let uri = req.uri().clone();
    //     let response = self
    //         .client
    //         .req(req)
    //         .await
    //         .map_err(ClientRequestError::RequestError)?
    //         .into_response_vec()
    //         .await?;
    //     <R>::parse_response(Some(request), &uri, response).map_err(Into::into)
    // }

    // /// Request on a valid [`RequestPut`] endpoint
    // pub async fn req_put<R, B, D, T>(
    //     &'a self,
    //     request: R,
    //     body: B,
    //     token: &T,
    // ) -> Result<Response<R, D>, ClientRequestError<<C as crate::HttpClient<'a>>::Error>>
    // where
    //     R: Request<Response = D> + Request + RequestPut<Body = B>,
    //     B: HelixRequestBody,
    //     D: serde::de::DeserializeOwned + PartialEq,
    //     T: TwitchToken + ?Sized,
    // {
    //     let req =
    //         request.create_request(body, token.token().secret(), token.client_id().as_str())?;
    //     let uri = req.uri().clone();
    //     let response = self
    //         .client
    //         .req(req)
    //         .await
    //         .map_err(ClientRequestError::RequestError)?
    //         .into_response_vec()
    //         .await?;
    //     <R>::parse_response(Some(request), &uri, response).map_err(Into::into)
    // }
}
