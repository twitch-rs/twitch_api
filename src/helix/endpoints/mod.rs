// pub mod bits;
// pub mod channels;
// #[cfg(feature = "unsupported")]
// pub mod charity;
// pub mod chat;
// pub mod clips;
// #[cfg(feature = "eventsub")]
// #[cfg_attr(nightly, doc(cfg(feature = "eventsub")))]
// pub mod eventsub;
// pub mod games;
// pub mod goals;
// pub mod hypetrain;
// pub mod moderation;
// pub mod points;
// pub mod polls;
// pub mod predictions;
// pub mod raids;
// pub mod schedule;
// pub mod search;
// pub mod streams;
// pub mod subscriptions;
// pub mod tags;
// pub mod teams;
pub mod users;
// pub mod videos;
// pub mod whispers;
