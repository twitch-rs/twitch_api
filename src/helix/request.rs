//! Requests for driving the API
pub mod errors;
use std::{convert::TryInto, str::FromStr};

use crate::parse_json;

use super::{ser, /* HelixRequestBody , */ HelixRequestError, InnerResponse, Response};
use errors::*;
/// A request is a Twitch endpoint, see [New Twitch API](https://dev.twitch.tv/docs/api/reference) reference
#[async_trait::async_trait]
pub trait Request: serde::Serialize {
    /// The path to the endpoint relative to the helix root. eg. `channels` for [Get Channel Information](https://dev.twitch.tv/docs/api/reference#get-channel-information)
    const PATH: &'static str;
    /// Scopes needed by this endpoint
    #[cfg(feature = "twitch_oauth2")]
    const SCOPE: &'static [twitch_oauth2::Scope];
    /// Optional scopes needed by this endpoint
    #[cfg(feature = "twitch_oauth2")]
    const OPT_SCOPE: &'static [twitch_oauth2::Scope] = &[];
    /// Response type. twitch's response will  deserialize to this.
    ///
    /// # Notes
    type Response: for<'y> yoke::Yokeable<'y>;

    /// Defines layout of the url parameters.
    fn query(&self) -> Result<String, errors::SerializeError> { ser::to_string(self) }
    /// Returns full URI for the request, including query parameters.
    fn get_uri(&self) -> Result<http::Uri, InvalidUri> {
        let query = self.query()?;
        let url = crate::TWITCH_HELIX_URL
            .join(<Self as Request>::PATH)
            .map(|mut u| {
                u.set_query(Some(&query));
                u
            })?;
        http::Uri::from_str(url.as_str()).map_err(Into::into)
    }
    /// Returns bare URI for the request, NOT including query parameters.
    fn get_bare_uri() -> Result<http::Uri, InvalidUri> {
        let url = crate::TWITCH_HELIX_URL.join(<Self as Request>::PATH)?;
        http::Uri::from_str(url.as_str()).map_err(Into::into)
    }
}

// /// Helix endpoint POSTs information
// pub trait RequestPost: Request {
//     /// Body parameters
//     type Body: HelixRequestBody;

//     /// Create a [`http::Request`] from this [`Request`] in your client
//     fn create_request(
//         &self,
//         body: Self::Body,
//         token: &str,
//         client_id: &str,
//     ) -> Result<http::Request<hyper::body::Bytes>, CreateRequestError> {
//         let uri = self.get_uri()?;
//
//        let body = body.try_to_body()?;
//        // eprintln!("\n\nbody is ------------ {:?} ------------", body);
//
//        let mut bearer = http::HeaderValue::from_str(&format!("Bearer {token}")).map_err(|_| {
//            CreateRequestError::Custom("Could not make token into headervalue".into())
//        })?;
//        bearer.set_sensitive(true);
//        http::Request::builder()
//            .method(http::Method::POST)
//            .uri(uri)
//            .header("Client-ID", client_id)
//            .header("Content-Type", "application/json")
//            .header(http::header::AUTHORIZATION, bearer)
//            .body(body)
//            .map_err(Into::into)
//    }
//
//     /// Parse response.
//     ///
//     /// # Notes
//     ///
//     /// Pass in the request to enable [pagination](Response::get_next) if supported.
//     fn parse_response<B: Into<hyper::body::Bytes>>(
//         // FIXME: Is this really needed? Its currently only used for error reporting.
//         request: Option<Self>,
//         uri: &http::Uri,
//         response: http::Response<B>,
//     ) -> Result<Response<Self, <Self as Request>::Response>, HelixRequestPostError>
//     where
//         Self: Sized,
//     {
//         let response: http::Response<hyper::body::Bytes> = response.map(|b| b.into());
//         let text = std::str::from_utf8(response.body().as_ref()).map_err(|e| {
//             HelixRequestPostError::Utf8Error(response.body().clone(), e, uri.clone())
//         })?;
//         if let Ok(HelixRequestError {
//             error,
//             status,
//             message,
//         }) = parse_json::<HelixRequestError>(text, false)
//         {
//             return Err(HelixRequestPostError::Error {
//                 error,
//                 status: status.try_into().unwrap_or(http::StatusCode::BAD_REQUEST),
//                 message,
//                 uri: uri.clone(),
//                 body: response.body().clone(),
//             });
//         }
//         <Self as RequestPost>::parse_inner_response(request, uri, text, response.status())
//     }

//     /// Parse a response string into the response.
//     fn parse_inner_response(
//         request: Option<Self>,
//         uri: &http::Uri,
//         response: &str,
//         status: http::StatusCode,
//     ) -> Result<Response<Self, <Self as Request>::Response>, HelixRequestPostError>
//     where
//         Self: Sized,
//     {
//         let response: InnerResponse<<Self as Request>::Response> = parse_json(response, true)
//             .map_err(|e| {
//                 HelixRequestPostError::DeserializeError(
//                     response.to_string(),
//                     e,
//                     uri.clone(),
//                     status,
//                 )
//             })?;
//         Ok(Response {
//             data: response.data,
//             pagination: response.pagination.cursor,
//             request,
//             total: response.total,
//             other: None,
//         })
//     }
// }

// /// Helix endpoint PATCHs information
// pub trait RequestPatch: Request {
//     /// Body parameters
//     type Body: HelixRequestBody;

//     /// Create a [`http::Request`] from this [`Request`] in your client
//     fn create_request(
//         &self,
//         body: Self::Body,
//         token: &str,
//         client_id: &str,
//     ) -> Result<http::Request<hyper::body::Bytes>, CreateRequestError> {
//         let uri = self.get_uri()?;

//         let body = body.try_to_body()?;
//         // eprintln!("\n\nbody is ------------ {} ------------", body);
//
//        let mut bearer = http::HeaderValue::from_str(&format!("Bearer {token}")).map_err(|_| {
//            CreateRequestError::Custom("Could not make token into headervalue".into())
//        })?;
//        bearer.set_sensitive(true);
//        http::Request::builder()
//            .method(http::Method::PATCH)
//            .uri(uri)
//            .header("Client-ID", client_id)
//            .header("Content-Type", "application/json")
//            .header(http::header::AUTHORIZATION, bearer)
//            .body(body)
//            .map_err(Into::into)
//    }
//
//     /// Parse response.
//     ///
//     /// # Notes
//     ///
//     /// Pass in the request to enable [pagination](Response::get_next) if supported.
//     fn parse_response<B: Into<hyper::body::Bytes>>(
//         // FIXME: Is this really needed? Its currently only used for error reporting.
//         request: Option<Self>,
//         uri: &http::Uri,
//         response: http::Response<B>,
//     ) -> Result<Response<Self, <Self as Request>::Response>, HelixRequestPatchError>
//     where
//         Self: Sized,
//     {
//         let response: http::Response<hyper::body::Bytes> = response.map(|b| b.into());
//         let text = std::str::from_utf8(response.body().as_ref()).map_err(|e| {
//             HelixRequestPatchError::Utf8Error(response.body().clone(), e, uri.clone())
//         })?;
//         if let Ok(HelixRequestError {
//             error,
//             status,
//             message,
//         }) = parse_json::<HelixRequestError>(text, false)
//         {
//             return Err(HelixRequestPatchError::Error {
//                 error,
//                 status: status.try_into().unwrap_or(http::StatusCode::BAD_REQUEST),
//                 message,
//                 uri: uri.clone(),
//                 body: response.body().clone(),
//             });
//         }
//         <Self as RequestPatch>::parse_inner_response(request, uri, text, response.status())
//     }

//     /// Parse a response string into the response.
//     fn parse_inner_response(
//         request: Option<Self>,
//         uri: &http::Uri,
//         response: &str,
//         status: http::StatusCode,
//     ) -> Result<Response<Self, <Self as Request>::Response>, HelixRequestPatchError>
//     where
//         Self: Sized;
// }

// /// Helix endpoint DELETEs information
// pub trait RequestDelete: Request {
//     /// Create a [`http::Request`] from this [`Request`] in your client
//     fn create_request(
//         &self,
//         token: &str,
//         client_id: &str,
//     ) -> Result<http::Request<hyper::body::Bytes>, CreateRequestError> {
//         let uri = self.get_uri()?;

//         let mut bearer = http::HeaderValue::from_str(&format!("Bearer {token}")).map_err(|_| {
//             CreateRequestError::Custom("Could not make token into headervalue".into())
//         })?;
//         bearer.set_sensitive(true);
//         http::Request::builder()
//             .method(http::Method::DELETE)
//             .uri(uri)
//             .header("Client-ID", client_id)
//             .header("Content-Type", "application/json")
//             .header(http::header::AUTHORIZATION, bearer)
//             .body(Vec::with_capacity(0).into())
//             .map_err(Into::into)
//     }
//     /// Parse response.
//     ///
//     /// # Notes
//     ///
//     /// Pass in the request to enable [pagination](Response::get_next) if supported.
//     fn parse_response<B: Into<hyper::body::Bytes>>(
//         // FIXME: Is this really needed? Its currently only used for error reporting.
//         request: Option<Self>,
//         uri: &http::Uri,
//         response: http::Response<B>,
//     ) -> Result<Response<Self, <Self as Request>::Response>, HelixRequestDeleteError>
//     where
//         Self: Sized,
//     {
//         let response: http::Response<hyper::body::Bytes> = response.map(|b| b.into());
//         let text = std::str::from_utf8(response.body().as_ref()).map_err(|e| {
//             HelixRequestDeleteError::Utf8Error(response.body().clone(), e, uri.clone())
//         })?;
//         if let Ok(HelixRequestError {
//             error,
//             status,
//             message,
//         }) = parse_json::<HelixRequestError>(text, false)
//         {
//             return Err(HelixRequestDeleteError::Error {
//                 error,
//                 status: status.try_into().unwrap_or(http::StatusCode::BAD_REQUEST),
//                 message,
//                 uri: uri.clone(),
//                 body: response.body().clone(),
//             });
//         }
//         <Self as RequestDelete>::parse_inner_response(request, uri, text, response.status())
//     }
//     /// Parse a response string into the response.
//     fn parse_inner_response(
//         request: Option<Self>,
//         uri: &http::Uri,
//         response: &str,
//         status: http::StatusCode,
//     ) -> Result<Response<Self, <Self as Request>::Response>, HelixRequestDeleteError>
//     where
//         Self: Sized;
// }

// /// Helix endpoint PUTs information
// pub trait RequestPut: Request {
//     /// Body parameters
//     type Body: HelixRequestBody;

//     /// Create a [`http::Request`] from this [`Request`] in your client
//     fn create_request(
//         &self,
//         body: Self::Body,
//         token: &str,
//         client_id: &str,
//     ) -> Result<http::Request<hyper::body::Bytes>, CreateRequestError> {
//         let uri = self.get_uri()?;

//         let body = body.try_to_body()?;
//         // eprintln!("\n\nbody is ------------ {} ------------", body);

//     let mut bearer = http::HeaderValue::from_str(&format!("Bearer {token}")).map_err(|_| {
//         CreateRequestError::Custom("Could not make token into headervalue".into())
//     })?;
//     bearer.set_sensitive(true);
//     http::Request::builder()
//         .method(http::Method::PUT)
//         .uri(uri)
//         .header("Client-ID", client_id)
//         .header("Content-Type", "application/json")
//         .header(http::header::AUTHORIZATION, bearer)
//         .body(body)
//         .map_err(Into::into)
// }

//     /// Parse response.
//     ///
//     /// # Notes
//     ///
//     /// Pass in the request to enable [pagination](Response::get_next) if supported.
//     fn parse_response<B: Into<hyper::body::Bytes>>(
//         // FIXME: Is this really needed? Its currently only used for error reporting.
//         request: Option<Self>,
//         uri: &http::Uri,
//         response: http::Response<B>,
//     ) -> Result<Response<Self, <Self as Request>::Response>, HelixRequestPutError>
//     where
//         Self: Sized,
//     {
//         let response: http::Response<hyper::body::Bytes> = response.map(|b| b.into());
//         let text = std::str::from_utf8(response.body().as_ref()).map_err(|e| {
//             HelixRequestPutError::Utf8Error(response.body().clone(), e, uri.clone())
//         })?;
//         if let Ok(HelixRequestError {
//             error,
//             status,
//             message,
//         }) = parse_json::<HelixRequestError>(text, false)
//         {
//             return Err(HelixRequestPutError::Error {
//                 error,
//                 status: status.try_into().unwrap_or(http::StatusCode::BAD_REQUEST),
//                 message,
//                 uri: uri.clone(),
//                 body: response.body().clone(),
//             });
//         }
//         <Self as RequestPut>::parse_inner_response(request, uri, text, response.status())
//     }

//     /// Parse a response string into the response.
//     fn parse_inner_response(
//         request: Option<Self>,
//         uri: &http::Uri,
//         response: &str,
//         status: http::StatusCode,
//     ) -> Result<Response<Self, <Self as Request>::Response>, HelixRequestPutError>
//     where
//         Self: Sized;
// }

/// Helix endpoint GETs information
pub trait RequestGet: Request
where for<'y> Self::Response: yoke::Yokeable<'y> {
    /// Create a [`http::Request`] from this [`Request`] in your client
    fn create_request(
        &self,
        token: &str,
        client_id: &str,
    ) -> Result<http::Request<hyper::body::Bytes>, CreateRequestError> {
        let uri = self.get_uri()?;

        let mut bearer = http::HeaderValue::from_str(&format!("Bearer {token}")).map_err(|_| {
            CreateRequestError::Custom("Could not make token into headervalue".into())
        })?;
        bearer.set_sensitive(true);
        http::Request::builder()
            .method(http::Method::GET)
            .uri(uri)
            .header("Client-ID", client_id)
            .header("Content-Type", "application/json")
            .header(http::header::AUTHORIZATION, bearer)
            .body(Vec::with_capacity(0).into())
            .map_err(Into::into)
    }

    /// Parse response.
    ///
    /// # Notes
    ///
    /// Pass in the request to enable [pagination](Response::get_next) if supported.
    fn parse_response<'de>(
        request: Option<Self>,
        uri: &http::Uri,
        response: &http::Response<&'de [u8]>,
    ) -> Result<Response<Self, <Self::Response as yoke::Yokeable<'de>>::Output>, HelixRequestGetError>
    where
        Self: Sized,
         yoke::trait_hack::YokeTraitHack<<Self::Response as yoke::Yokeable<'de>>::Output>:
            serde::Deserialize<'de>,
    {
        let text = std::str::from_utf8(response.body()).map_err(|e| {
            HelixRequestGetError::Utf8Error(response.body().to_vec(), e, uri.clone())
        })?;
        //eprintln!("\n\nmessage is ------------ {} ------------", text);
        if let Ok(HelixRequestError {
            error,
            status,
            message,
        }) = parse_json::<HelixRequestError>(text, false)
        {
            return Err(HelixRequestGetError::Error {
                error,
                status: status.try_into().unwrap_or(http::StatusCode::BAD_REQUEST),
                message,
                uri: uri.clone(),
            });
        }
        <Self as RequestGet>::parse_inner_response(request, uri, text, response.status())
    }

    /// Parse a response string into the response.
    fn parse_inner_response<'de>(
        request: Option<Self>,
        uri: &http::Uri,
        response: &'de str,
        status: http::StatusCode,
    ) -> Result<Response<Self, <Self::Response as yoke::Yokeable<'de>>::Output>, HelixRequestGetError>
    where
        Self: Sized,
         yoke::trait_hack::YokeTraitHack<<Self::Response as yoke::Yokeable<'de>>::Output>:
            serde::Deserialize<'de>,
    {
        let response: InnerResponse<
            yoke::trait_hack::YokeTraitHack<<Self::Response as yoke::Yokeable<'de>>::Output>,
        > = parse_json(response, true).map_err(|e| {
            HelixRequestGetError::DeserializeError(response.to_string(), e, uri.clone(), status)
        })?;
        Ok(Response {
            data: response.data.0,
            pagination: response.pagination.cursor,
            request,
            total: response.total,
            other: response.other,
        })
    }
}
