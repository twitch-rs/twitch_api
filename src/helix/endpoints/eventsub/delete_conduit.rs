//! Deletes a specified conduit.
//! [`delete-conduit`](https://dev.twitch.tv/docs/api/reference/#delete-conduit)
//!
//! Note that it may take some time for Eventsub subscriptions on a deleted conduit to show as disabled when calling [Get Eventsub Subscriptions](super::get_eventsub_subscriptions).

use super::*;
use helix::RequestDelete;

/// Query Parameters for [Delete Conduit](super::delete_conduit)
///
/// [`delete-conduit`](https://dev.twitch.tv/docs/api/reference/#delete-conduit)
#[derive(PartialEq, Eq, Serialize, Clone, Debug)]
#[cfg_attr(feature = "typed-builder", derive(typed_builder::TypedBuilder))]
#[must_use]
#[non_exhaustive]
pub struct DeleteConduitRequest<'a> {
    /// Conduit ID.
    #[cfg_attr(feature = "typed-builder", builder(setter(into)))]
    #[cfg_attr(feature = "deser_borrow", serde(borrow = "'a"))]
    pub id: Cow<'a, types::ConduitIdRef>,
}

impl<'a> DeleteConduitRequest<'a> {
    /// Delete a specific conduit
    pub fn new(id: impl types::IntoCow<'a, types::ConduitIdRef> + 'a) -> Self {
        Self { id: id.into_cow() }
    }
}

/// Return Values for [Delete Conduit](super::delete_conduit)
///
/// [`delete-conduit`](https://dev.twitch.tv/docs/api/reference/#delete-conduit)
#[derive(PartialEq, Eq, Deserialize, Serialize, Debug, Clone)]
#[non_exhaustive]
pub enum DeleteConduitResponse {
    /// 204 - Conduit deleted
    Success,
}

impl Request for DeleteConduitRequest<'_> {
    type Response = DeleteConduitResponse;

    const PATH: &'static str = "eventsub/conduits";
    #[cfg(feature = "twitch_oauth2")]
    const SCOPE: twitch_oauth2::Validator = twitch_oauth2::validator![];
}

impl RequestDelete for DeleteConduitRequest<'_> {
    fn parse_inner_response(
        request: Option<Self>,
        uri: &http::Uri,
        response: &str,
        status: http::StatusCode,
    ) -> Result<helix::Response<Self, Self::Response>, helix::HelixRequestDeleteError>
    where
        Self: Sized,
    {
        match status {
            http::StatusCode::NO_CONTENT | http::StatusCode::OK => Ok(helix::Response::with_data(
                DeleteConduitResponse::Success,
                request,
            )),
            _ => Err(helix::HelixRequestDeleteError::InvalidResponse {
                reason: "unexpected status",
                response: response.to_string(),
                status,
                uri: uri.clone(),
            }),
        }
    }
}

#[cfg(test)]
#[test]
fn test_response() {
    helix::assert_helix_snapshot!(
        DeleteConduitRequest:
        req = DeleteConduitRequest::new("bfcfc993-26b1-b876-44d9-afe75a379dac"),
        @r#"
    uri
    ----
    https://api.twitch.tv/helix/eventsub/conduits?id=bfcfc993-26b1-b876-44d9-afe75a379dac

    response
    ----
    Response {
        data: Success,
        pagination: None,
        request: Some(
            DeleteConduitRequest {
                id: "bfcfc993-26b1-b876-44d9-afe75a379dac",
            },
        ),
        total: None,
        other: None,
    }
    "#,
    );
}
