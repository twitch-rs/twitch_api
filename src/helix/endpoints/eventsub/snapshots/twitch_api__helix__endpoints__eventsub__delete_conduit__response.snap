---
source: src/helix/endpoints/eventsub/delete_conduit.rs
expression: uri
---
https://api.twitch.tv/helix/eventsub/conduits?id=bfcfc993-26b1-b876-44d9-afe75a379dac
